<a href="https://colab.research.google.com/github/andrearivera019/Analitica-de-Negocios/blob/main/Reto_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Caso De Estudio: Detección de Riesgo SARLAFT en una Entidad Financiera**

En este documento se desarrolla y analiza un modelo Naive Bayes para la detección de riesgo SARLAFT (lavado de activos) en una entidad financiera utilizando diferentes variables socioeconómicas y basándose en el comportamiento financiero y los  cambios inusuales en su patrimonio; para de esta manera determinar si las personas tienen un alto riesgo de incurrir en lavado de activos o no lo tienen.

Esto se realiza debido a que la entidad a pesar de contar con unos controles tradicionales de monitoreo, detectó que se presentaron operaciones sospechosas que no se identificaron de manera oportuna.


De aucerdo con lo anterior, las variables tomadas en cuenta se explican a continuación:

**Edad:** es la edad de la persona en el momento en que se está realizando su evaluación para saber si incurre en lavado de actvivos o no.

**Ingresos:** son los ingresos recibidos por una persona como resultado de una actividad comercial.

**Egresos:** son los gastos que tiene cada una de las personas (se tienen en cuenta gastos de todo tipo).

**Monto Transado:** es el monto que las personas movieron en sus tarjetas de crédito durante el periodo de tiempo que se realizó el seguimiento a las transacciones.

**Número de Tarjetas de Crédito:** cantidad de tarjetas de crédito que posee la persona al momento de la evaluación.

**Porcentaje Crecimiento Patrimonio:** es el crecimiento del patrimonio expresado en porcentaje, para analizar cuánto aumentó o disminuyó con respecto a la anterior información. Entendiendo el patrimonio como el conjunto total de sus bienes (materiales e inmateriales) valorables económicamente al momento de la evaluación.

Cabe aclarar que todas las variables están expresadas en USD, exceptuando la edad, el porcentaje de crecimiento del patrimonio y el número de tarjetas de crédito.


***0. Se procede con la carga de las librerías de trabajo***



In [ ]:
import numpy as np
import pandas as pd

#Librerías específicas
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


***1. Se cargan los datos de trabajo de la base de datos LavadoActivos***

In [ ]:
nxl = '/content/drive/MyDrive/Analítica de Negocios/2. LavadoActivos.xlsx'
XDB = pd.read_excel(nxl,sheet_name=0)
XDB.head(100)

#Seleccionamos variables de trabajo
XD = XDB [['edad','ingresos','gastos','monto transado', 'tarjetas credito', 'porcentaje_crecimiento_patrimonio']]
XD.head(10)
yd = XDB[['lavado_activos']] #Variable de decisión
yd.head()


,lavado_activos
0,1
1,1
2,0
3,0
4,0


***2. Se implementa el Modelo Naive Bayes***

In [ ]:
np.set_printoptions(suppress=True,precision=3) #3 decimales
mnb = GaussianNB()
mnb.fit(XD,yd) #Ajustar variables Entrada - Salida

#Mostramos las medidas de las variables
u = mnb.theta_
sigma = mnb.var_ ; sigma=np.sqrt(sigma)
print("'edad','ingresos','gastos','monto transado', 'tarjetas credito', porcentaje_crecimiento_patrimonio'")
print(u)
print()
print()
print("Las desviaciones son: ")
print(sigma)
print()
print()
print("Los límites superiores de las variables son: ",u+sigma)
print()
print()
print("Los límites inferiores de las variables son: ",u-sigma)
print()
print()


'edad','ingresos','gastos','monto transado', 'tarjetas credito', porcentaje_crecimiento_patrimonio'
[[   45.47   6690.733  3601.75   3090.464     1.955    44.687]
 [   44.3   70262.304 38707.284 33835.62      2.016    44.821]]


Las desviaciones son: 
[[    14.711   6206.964   3219.016   2456.859      2.535     14.173]
 [    15.052 113412.265  65902.557  62991.391      2.557     15.479]]


Los límites superiores de las variables son:  [[    60.18   12897.697   6820.765   5547.323      4.49      58.86 ]
 [    59.352 183674.569 104609.842  96827.011      4.574     60.299]]


Los límites inferiores de las variables son:  [[    30.759    483.768    382.734    633.605     -0.581     30.514]
 [    29.248 -43149.96  -27195.273 -29155.771     -0.541     29.342]]




/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


#**Análisis de Resultados: Implementación del Modelo Naive Bayes**

De los resultados dados por el modelo, se puede evidenciar que la categoría de las personas que tienen un alto riesgo de incurrir en lavado de activos cuentan con una edad menor a las personas que no tienen un alto riesgo de incurrir en el lavado de activos, mostrando unos datos de 44.3 y 45.47 respectivamente; esto nos indica que los perfiles con menor edad tienen con mayor frecuencia características asociadas a alertas de riesgo, y esto podría estar relacionado con situaciones como menores niveles de estabilidad financiera; no obstante, esta diferencia no implica una relación causal directa entre la edad y el riesgo SARLAFT, sino que es una asociación dentro de la muestra analizada.

De igual manera, se puede evidenciar que si bien la media de las personas que incurren en lavado de activos tienen una media de 44 y esta puede subir hasta 59 y bajar hasta 29 (es decir, oscila entre esas edades); y para las personas que no incurren en lavado de activos, tienen una media de 45 y puede subir hasta 60 y bajar hasta 30 (es decir, oscila entre esas edades).

Por otro lado, con respecto a los ingresos y a los gastos de estas personas, se puede evidenciar que para el caso de ambos, las personas que no incurren en lavado de activos tienen tanto ingresos como gastos menores que las personas que si incurren en lavado de activos; sin embargo la presencia simultánea de ingresos y gastos elevados no constituye por sí sola una conducta ilícita, pero si se combina con otros factores que se evalúan en el modelo, esto se convierte en un elemento importante para identificar alertas de SARLAFT. Por ejemplo, si se combina con el monto transado (que también fue una variable que se analizó), se evidencia que este también es mayor en las personas que son opcionadas para incurrir en lavado de activos que en las que no son opcionadas, y esto combinado con los altos ingresos y gastos, sugiere que hay una alta rotación y volumen de capital (en este caso el monto transado de las personas que incurren en lavado de activos es de 33835 aprox y de las personas que no incurren de 3090 aprox).

Se puede observar que tanto el número promedio de tarjetas de crédito como el porcentaje de crecimiento del patrimonio presentan valores muy similares entre los clientes que son opcionados y no opcionados en incurrir en lavado de activos; y se puede deducir que estas variables no actúan como factores determinantes en la identificación del riesgo de lavado de activos, dándonos una idea de que el riesgo se manifiesta principalmente en la dinámica y volumen de las transacciones, más que en la acumulación patrimonial o en el acceso que tengan a productos financieros; es decir, el riesgo no está en cuánto patrimonio tienen ni en cuántas tarjetas de crédito tienen, sino en cómo y en cuánto dinero se mueve.


***3. Continuamos con la evaluación del Modelo, por medio de la matriz de confusión***

In [ ]:
ydp = mnb.predict(XD)
cm = confusion_matrix(yd,ydp)
print(cm)
print()

VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]; TDatos=len(XDB)

#1. Exactitud: funcionamiento general del modelo
Ex=(VP+VN)/TDatos
print("Exactitud: ",Ex)
print()

#2. Tasa Error: porcentaje de fallos del modelo
TEr=(FP+FN)/TDatos
print("Tasa Error: ",TEr)
print()

#3. Sensibilidad: como se comportó con respecto a los positivos
Se=VP/(VP+FN)
print("Sensibilidad: ",Se)
print()

#4. Especificidad: como se comporta pronosticando negativos
Es=VN/(VN+FP)
print("Especificidad: ",Es)
print()

#5. Precisión: es una versión de cómo se comporta el modelo frente a los positivos solamente
Pr=VP/(VP+FP)
print("Precisión: ",Pr)
print()

#6. Predicción Negativa:
PrN=VN/(VN+FN)
print("Predicción Negativa: ",PrN)
print()


[[2179   44]
 [ 145  782]]

Exactitud:  0.94

Tasa Error:  0.06

Sensibilidad:  0.8435814455231931

Especificidad:  0.9802069275753487

Precisión:  0.9467312348668281

Predicción Negativa:  0.9376075731497419



#**Análisis de Resultados: Evaluación del Modelo y la Matriz de Confusión**

Se presenta la matriz de confusión, la cual nos permite comparar las predicciones del modelo con los valores reales, mostrándonos cuáles de los casos fueron clasificados correctamente y cuántos de manera incorrecta; en este caso, se obtuvo que el modelo identifica a 2179 verdaderos negaticos (personas sin riesgo correctamente clasificadas por el modelo), 44 falsos positivos (personas sin riesgo que fueron clasificadas erróneamente como riesgosas; 145 falsos negativos (personas con riesgo que el modelo no logró detectar) y 782 verdaderos positivos (personas con riesgo de lavado de activos correctamente identificadas).

**Exactitud**: esta representa la proporción total de observaciones que fueron clasificadas correctamente por el modelo, para este caso, nos dio una exactitud del 94% aproximadamente, un valor que es bastante alto, lo que nos indica que el modelo tiene un desempeño muy alto y que logra clasificar correctamente la mayoría de los casos que analiza.

**Tasa de Error:** nos indica el porcentaje de observaciones clasificadas incorrectamente, y en el caso de este modelo, tenemos una tasa de error de aproximadamente el 6%, porcentaje que es muy bajo y refuerza la solidez del modelo.

**Sensibilidad:** mide la capacidad del modelo para identificar de manera correcta a las personas que sí presentan un riesgo de incurrir en lavado de activos. Para este modelo, tenemos una sensibilidad del 84% aproximadamente, lo que indica que el modelo alcanza a detectar en gran medida los casos de riesgo y esto reduce significativamente la cantidad de personas con riesgo que no son identificadas.

**Especificidad:** esta mide la capacidad del modelo para identificar correctamente a las personas que no presentan riesgos, y en el caso de nuestro modelo, tenemos que es del 98% aproximadamente, valor que es significativamente aglo y nos dice que el modelo tiene un desempeño muy bueno y tiende a minimizar la generación de alertas innecesarias.

**Precisión:** representa la proporcipon de las personas clasificadas como riesgosas que realmente lo son; para este caso nos dio una precisión del 94% aproximadamente y esto significa que casi todas las alertas generadas por el modelo son acertadas, y cabe aclarar que es un porcentaje bastante alto que nos dice que el modelo es eficaz y bueno.

**Predicción negativa:** representa la proporción de personas clasificadas como no riesgosas que relamente no presentan riesgo y un valor del 93% aproximadamente es bastante bueno, indicando una alta confiabilidad en las decisiones de descarte que realice el modelo.

Finalmente y a modo de sintetizar, se puede decir que el modelo tiene un desempeño bastante bueno y con una exactitud, precisión, especificidad y una sensibilidad bastante altas que permiten una adecuada identificación de casos con riesgo de lavado de activos, manteniendo un nivel bajo de errores.

***4. Evaluar al cliente***

In [ ]:
XP = [34,9413,5851,2,7096,37.85]
ydc = mnb.predict([XP])
print(ydc)
print()

if ydc==1:
  print("La persona analizada tiene un alto riesgo de incurrir en lavado de activos")
else:
  print("La persona analizada no tiene un riesgo alto de incurrir en lavado de activos")

print()

[1]

La persona analizada tiene un alto riesgo de incurrir en lavado de activos



/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


#**Análisis de Resultados: Evaluación de un Solicitante Específico**

Al momento de evaluar al solicitante mediante el modelo de Naive Bayes, pudimos obtener una clasificación que corresponde a un alto riesgo de incurrir en lavado de activos, y esto se explica principalmente por la combinación de una edad que es relativamente menor y la presencia de ingresos, gastos y montos transados elevados representando alta rotación y volumen de capital, y confirmándonos además lo que decíamos anteriormente de que estas son las variables decisivas para determinar si incurren o no en el lavado de activos.

Si bien las variables como el número de las tarjetas de crédito y el crecimiento del patrimonio no son determinantes, el comportamiento transaccional del solicitante si genera alertas relevantes.

